modeling/algorithm/local-weighted-slice-algorithm.go

```diff
@@ -0,0 +1,211 @@
+		}
+		sliceGroups[zoneName] = localGroup
+
+		// push 0.xx deviation into corresponding lists
+		// 2.3 expected endpoints, 3 ownd endpoints, 0.7 decimal deviation, 0.3
+		// endpoints and 2 endpoints needed
+		// 2.4 expected endpoints, 1 owned endpoint, -0.4 decimal deviation, 0.4
+		// endpoints and 2 endpoints needed
+		decimalDeviation := deviation - float64(intDeviation)
+		if decimalDeviation > 0 {
+			// One endpoint from this zone will contribute 1-decimalDeviation to
+			// the local zone, and contribute remaining to other zones. This
+			// will be implemented through routing weights of EndpointSliceGroup
+			weightedEndpointsAvailable.push(endpointDeviation{name: zoneName, deviation: 1, weight: 1 - decimalDeviation, consumeByLocal: true})
+		} else if decimalDeviation < 0 {
+			weightedEndpointsNeeded.push(endpointDeviation{name: zoneName, deviation: 1, weight: -decimalDeviation})
+		}
+	}
+
+	err := alg.balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, &weightedEndpointsAvailable, &weightedEndpointsNeeded, sliceGroups)
+	return sliceGroups, err
+}
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalWeightedSliceAlgorithm) balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, weightedEndpointsAvailable *endpointsList, weightedEndpointsNeeded *endpointsList, sliceGroups map[string]types.EndpointSliceGroup) error {
+
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// There are no more full endpoints available, but this zone still needs
+		// endpoints. Push the needed endpoints into a weighted list and deal
+		// with them as partial endpoints.
+		if len(endpointsAvailable.byZone) == 0 {
+			receiveZone.weight = 1
+			weightedEndpointsNeeded.push(receiveZone)
+			endpointsNeeded.pop()
+			continue
+		}
+		// same logic as original local-slice algorithm
+		for index := 0; index < len(endpointsAvailable.byZone); {
+			sendZone := endpointsAvailable.byZone[index]
+			if sendZone.deviation == receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: sendZone.deviation, Weight: 1}
+				receiveZone.deviation = 0
+				endpointsAvailable.pop()
+				break
+			}
+			if sendZone.deviation > receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: receiveZone.deviation, Weight: 1}
+				endpointsAvailable.byZone[index].deviation -= receiveZone.deviation
+				receiveZone.deviation = 0
+				break
+			}
+			if sendZone.deviation < receiveZone.deviation {
+				sliceGroups[receiveZone.name].Composition[sendZone.name] = types.WeightedEndpoints{Number: sendZone.deviation, Weight: 1}
+				receiveZone.deviation -= sendZone.deviation
+				endpointsAvailable.pop()
+				continue
+			}
+		}
+		// if needed.deviation != 0 means more full endpoints needed than
```

Comment:
When would this happen? It seems strange to equate != 0 meaning more needed than available, wouldn't a negative value have a different meaning than a positive value here?

Reply:
changed to > 0 and added more comments to explain